core/pubnub_subscribe_event_engine.c

```diff
@@ -0,0 +1,802 @@
+
+    return set;
+}
+
+pubnub_subscription_set_t*
+pubnub_subscription_set_alloc_with_subscriptions(
+    pubnub_subscription_t*               sub1,
+    pubnub_subscription_t*               sub2,
+    const pubnub_subscription_options_t* options)
+{
+    PUBNUB_ASSERT_OPT(NULL != sub1);
+    PUBNUB_ASSERT_OPT(NULL != sub2);
+
+    pubnub_subscription_set_t* set = _subscription_set_alloc(
+        options,
+        SUBSCRIPTIONS_LENGTH);
+    if (NULL == set) { return NULL; }
+
+    if (PNR_OUT_OF_MEMORY == pubnub_subscription_set_add(set, sub1) ||
+        PNR_OUT_OF_MEMORY == pubnub_subscription_set_add(set, sub2)) {
+        pubnub_subscription_set_free(set);
+        return NULL;
+    }
+
+    set->ee = sub1->ee;
+
+    return set;
+}
+
+enum pubnub_res pubnub_subscription_set_add(
+    pubnub_subscription_set_t* set,
+    pubnub_subscription_t*     sub)
+{
+    PUBNUB_ASSERT_OPT(NULL != set);
+    PUBNUB_ASSERT_OPT(NULL != sub);
+
+    pubnub_mutex_lock(set->mutw);
+    const bool subscribed = set->subscribed;
+
+    if (pbhash_set_contains(set->subscriptions, sub->entity->id.ptr)) {
+        PUBNUB_LOG_INFO(
+            "pubnub_subscription_set_add: subscription (sub=%p) for entity "
+            "('%s') already added into set",
+            sub,
+            sub->entity->id.ptr);
+        pubnub_mutex_unlock(set->mutw);
+        return PNR_SUB_ALREADY_ADDED;
+
+    }
+
+    if (PBHSR_OUT_OF_MEMORY == pbhash_set_add(set->subscriptions,
+                                              sub->entity->id.ptr,
+                                              sub)) {
+        pubnub_mutex_unlock(set->mutw);
+        PUBNUB_LOG_ERROR(
+            "pubnub_subscription_set_add: failed to allocate memory "
+            "for subscription\n");
+        return PNR_OUT_OF_MEMORY;
+    }
+
+    _subscription_reference_count_update(sub, true);
+    pubnub_mutex_unlock(set->mutw);
+
+    if (subscribed) {
+        // Notify changes in active subscription set.
+        pbcc_subscribe_ee_change_subscription_with_subscription_set(
+            set->ee,
+            set,
+            sub,
+            true);
+    }
+
+    return PNR_OK;
+}
+
+enum pubnub_res pubnub_subscription_set_remove(
+    pubnub_subscription_set_t* set,
+    pubnub_subscription_t*     sub)
+{
+    PUBNUB_ASSERT_OPT(NULL != set);
+    PUBNUB_ASSERT_OPT(NULL != sub);
+
+    pubnub_mutex_lock(set->mutw);
+    const bool subscribed = set->subscribed;
+
+    if (!pbhash_set_contains(set->subscriptions, sub->entity->id.ptr)) {
+        PUBNUB_LOG_INFO(
+            "pubnub_subscription_set_remove: subscription (sub=%p) for entity "
+            "('%s') not found",
+            sub,
+            sub->entity->id.ptr);
+        pubnub_mutex_unlock(set->mutw);
+        return PNR_SUB_NOT_FOUND;
+    }
+
+    if (subscribed) {
+        // Temporarily release lock to avoid deadlocks from subscription ee core
+        // code access the list of subscribables.
+        pubnub_mutex_unlock(set->mutw);
```

Comment:
I'm a little bit worried about this statement. 
Is it possible that this quite short amount of time unguarded mutex might mess anything in execution?